### DISCUSSION TITLE
MPF 0.56 doesn't work with P(3)-Roc

### DISCUSSION THREAD

**Discussion by Midway-Man** [missionpinball/missionpinball#80] [Q&A]
Just updated my machine from MPF0.55 Python3.8 to MPF0.56 Python3.9 

But unfortunately my game no longer works. It crashes at startup. No changes have been made to the code itself. 

When i start MPF with virtual hardware -X my code runs fine. The issue must be related to one of the new versions of MPF or Python. Unfortunately I am not an expert, so it would be nice if one of the DEVs could help my out. 

That's part of the log file. 

2023-02-17 12:48:39,148 : INFO : Machine : Initialise MPF.
2023-02-17 12:48:39,221 : ERROR : Machine : Failed to initialise MPF
Traceback (most recent call last):
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\platforms\p_roc_common.py", line 31, in <module>
    import pinproc
ModuleNotFoundError: No module named 'pinproc'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\platforms\p_roc_common.py", line 39, in <module>
    from mpf.platforms.pinproc.windows import pinproc
ImportError: cannot import name 'pinproc' from 'mpf.platforms.pinproc.windows' (unknown location)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\core\machine.py", line 679, in initialise_mpf
    raise init.exception()
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\core\machine.py", line 246, in initialise
    await self.initialise_core_and_hardware()
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\core\machine.py", line 232, in initialise_core_and_hardware
    self._load_hardware_platforms()
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\core\machine.py", line 479, in _load_hardware_platforms
    self.add_platform(hardware_platform)
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\core\machine.py", line 613, in add_platform
    self.hardware_platforms[name] = hardware_platform(self)
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\platforms\p_roc.py", line 44, in __init__
    super().__init__(machine)
  File "C:\Users\Pinball\.local\pipx\venvs\mpf\lib\site-packages\mpf\platforms\p_roc_common.py", line 180, in __init__
    raise MpfRuntimeError('Could not import "pinproc". Either the library is not installed or is missing '
mpf.exceptions.runtime_error.MpfRuntimeError: Runtime Error in P-Roc: Could not import "pinproc". Either the library is not installed or is missing some of its dependencies. Check the install instructions for your OS in Multimorphic section of the MPF docs. You can run mpf with "-X" to use virtual hardware in the meantime. Error Code: RE-P-Roc-3 (https://docs.missionpinball.org/en/0.56/logs/RE-P-Roc-3.html)

Thanks in advance for helping us!

**Comment by atummons** [comment]
Try following the steps here: https://docs.missionpinball.org/en/stable/hardware/multimorphic/hardware_drivers.html

Shot in the dark, but since you upgraded python, do you need to reinstall the multimorphic stuff again.

**Reply by Midway-Man** [reply]
Gave it a try, but the issue still persists. Windows Device Manager does show the P3-Roc as connected.  I still suspect MPF or Python does cause this problem. 

Is there a way to downgrade to MPF0.55? Until the Devs resolve the current issue.

**Comment by qcapen** [comment]
MPF has always used a special version of the p-roc drivers on Windows.
Unfortunately, Multimorphic has never supported or created a Python 3
compatible driver library. One was created specifically for MPF a long time
ago. The problem has been building that driver for Windows. The way it is
now, compiling and building the p-roc driver has been a difficult manual
process for each new version of Python. I managed to do this for the
project a few years ago when Python 3.6 was released, but the project
didn't work for Python 3.7+ and I haven't had any time to try to figure it
out (and I don't own any p-roc or p3-roc boards to test it). MPF (or really
someone connected to Multimorphic) needs to maintain a set of Python 3
compatible drivers (or set something up using github actions to
automatically build).

Bottom line, the problem is not specifically with MPF 0.55 but with not
having the Windows p-roc drivers for newer versions of Python 3 (and not
having anyone to create and maintain them).

Quinn

**Reply by Midway-Man** [reply]
Thanks for the detailed reply. As far as I understand, the issue seems to coming from Python 3.9. Not MPF0.56.
Creating a Windows compatible driver for P-Roc sounds like a more complicated task. I guess it can take weeks if not months until this issue is sorted out. 
Would be good if someone could add a Warning to the MPF0.56 install instructions document, stating that P-Roc is currently not supported on Windows platforms. To prevent more people from having this problem, just until it is fixed. 

Plan A: Unfortunately the old documents seem to be gone, is there a way to downgrade to MPF0.55?

Plan B: I was thinking about moving to Linux anyway, since my hardware setup is not the strongest and Windows uses unnecessary resources. Linux on the other hand seems to be a less power hungry system. Also at one point I wanted to create a read only system anyway which seems to be a no-go with Windows. Du you think P-Roc would work on Linux with current MPF and Python? Without needing complicated hardware drivers?

What do you think?
Thanks in advance for the support!

**Reply by atummons** [reply]
@qcapen Do you have any of the precursor files that created the previous versions?  All I can see is the files, but since they are machine language I have no idea if one value needs changed to support 3.9 or if it needs a million things changed.  Gerry's comment was if it requires a version that isn't there it would be "unfortunate".  So I doubt we will be getting any help from them.

If you can send me the raw files, I can try to work them for the later versions of python (if I can keep up with what is going on).  Otherwise, I feel we need make specific notes that P3-Roc hardware is not supported for newer versions/on specific OS to save new entrants from running into this issue.

Thanks!

**Reply by qcapen** [reply]
I looked back in my dev folder for the p-roc libraries in order to try to remember how I built the pypinproc library years ago. My folder contains 2GB of source code, libraries, versions for 32-bit, 64-bit, etc. It's unfortunately kind of a mess so I'm not sure what exactly to try to send you. Older versions of Python required different versions of Visual Studio to build and link to the Python API. The basic process was to install the Python APIs for the target Python version (both 32-bit and 64-bit), download the libpinproc source code from github (https://github.com/preble/libpinproc), download the customized pypinproc code that supports python 3 from github (https://github.com/missionpinball/pypinproc), download the usb and ftdi libraries, and then experiment using visual studio to get the whole thing built and compiled. What made it more challenging is I've never owned any p-roc boards to do any testing. Once I got it working it wasn't hard to support several versions of Python. However, I could never get it to work in Visual Studio 2019 to support the newer Python versions (could never solve several linking errors). pypinproc is basically a python interface wrapper around the libpinproc library (all written in C).

It looks like libpinproc has been updated since the last time I tried building it and now includes instructions for VS2019 so that might be a good place to start. My intention was to abandon the local build process and attempt to setup github actions to convert the build process to a CI pipline which would be much easier to maintain, but I got too busy to give it a try (and not owning any p-roc hardware didn't help my motivation). 

I'm happy to try to support the effort to get this solved and can provide my old VS project from VS2017 that supported python 3.6, but I am very limited on time right now. Let me know what I can do. I am certainly no build expert, but was able to get it figured out in the past.

Quinn

**Reply by atummons** [reply]
Quinn, Can you help me out just a little more here?  I now have VS 2019, and the python stuff there.  I have libpinproc and pypinproc source code.  I got the FTDI drivers, not sure what you mean by the USB drivers.

You say, get the whole thing built and compiled.  Is that running the setup.py file in pypinproc and going line by line?  Do I need to actually install libpinproc on my machine first?  I think two steps further, and I can hopefully understand what I need to do to attempt this process.

Then once you are done, is this what generated the pinproc.cp38-win_amd64.pyd files? Or does it do something else and then you create that file?

I know you are busy, and no rush, and sorry to bother!

**Reply by Midway-Man** [reply]
Any news on that? :)

**Reply by atummons** [reply]
@qcapen Tagging, as I am not sure you got the other one.  If you did and are busy, then feel free to ignore! :)

**Reply by qcapen** [reply]
I did get the last one and spent some time trying to do another build, but failed and I have not had time to pursue it any further (and forgot to reply back). Here is what my plan was:

1. Follow the instructions [here](https://github.com/preble/libpinproc#building-in-windows-with-cmake-and-visual-studio-2019) to build libpinproc with Visual Studio 2019. Build it as a library to link to pypinproc.
2. Download the Mission Pinball version of the pypinproc code [here](https://github.com/missionpinball/pypinproc). Build it and link it to whatever version of Python desired (or multiple versions). 

As an alternative I was going to try to figure out GitHub Actions to build it automatically on the Mission Pinball pypinproc GitHub repository (I don't have any experience with GitHub Actions so this would definitely take some time).

The output of the process is the pinproc.cp38-win_amd64.pyd files (or whatever Python version you build and link to). Ideally it would be nice to automate in GitHub where it could more easily be maintained. 

Unfortunately I didn't even get past step 1 before running into linking issues. I think it was FTDI related, but I don't remember. My job (software engineer) has kept me extremely busy lately so I haven't been able to find the time to devote to pinball and jumping back on the computer after a long day has not been at the top of my list. Hopefully things will calm down later in September and I'll have some time and energy to try this again. This really shouldn't be this hard, but building on Windows has never been that easy with Python. I'm still unsure why the audio library appears to build fine on Windows but doesn't produce any sound.

Quinn

**Reply by qcapen** [reply]
Some possible good news. I decided to give this another go this afternoon and I think I've been successful in getting a 64-bit Windows version built with Python 3.9. Here's what I did:

1. Download the Mission Pinball version of the pypinproc source code: [](https://github.com/missionpinball/pypinproc)
2. Unzip the pypinproc code into a working directory (I called it pypinproc)
3. Download the FTDI USB drivers from [](https://ftdichip.com/drivers/d2xx-drivers/) (I downloaded the 64-bit Windows drivers: [](https://www.ftdichip.com/Drivers/CDM/CDM%20v2.12.36.4%20WHQL%20Certified.zip) )
4. Unzip the FTDI drivers into a sub-directory of pypinproc (I used ftd2xx)
5. Download the libpinproc source code: [](https://github.com/preble/libpinproc/tree/dev)
6. Unzip the libpinproc source code into a subdirectory of pypinproc (I used libpinproc)
7. Create a Visual Studio C++ project for a 64-bit Windows dll (I created an empty project)
8. Include all the .cpp and .h files to the project in the pypinproc folder, py3c subfolder, ftd2xx subfolder (ftd2xx.h), libpinproc\src subfolder, libpinproc\include subfolder
9. Add additional include paths to the project: py3c subfolder, ftd2xx subfolder, libpinproc\include subfolder, and the Python include directory (where Python.h is located for the version of python you want to build the library for; Python 3.9 in my case)
10. Add the following additional libraries to the project linking section: ftd2xx.lib, Python3.9.lib
11. Add the following addition library paths to the project: ftd2xx\amd64 (where ftd2xx.lib is located), Python libs path (where Python3.9.lib is located)
12. I had to change some VC++ project options to get things to compile (add /permissive to Additional Options in C/C++ All Options, set SDL Checks option to No). I think that was it.
13. Build the dll

I cannot test it as I don't have any P-Roc hardware, but it built successfully on my system. Other versions of Python should be able to be compiled just by changing the linked Python library and changing the Python include and libs directories.

**Reply by atummons** [reply]
You're awesome! I'll try to follow those steps as well to make sure I can do it for my own knowledge and future support.

I could upgrade mine to run 3.9 on windows with P3-ROC hardware. What are the risks? Does it either work or not work and will error out? Or any risk to "beta" a new version from the hardware perspective?

**Reply by qcapen** [reply]
Shouldn't be any risk. It will either work or it won't. Also, you'll have to make sure the ftdi dll is either in the path or in the same directory as the pypinproc dll. I seem to remember that the ftd2xx.dll file has the wrong name in the ftdi zip file (I think it is ftd2xx64.dll and needs to be renamed to ftd2xx.dll). Also, I used the default VC++ runtime version with Visual Studio 2019, but some versions of Python may require different Visual C++ runtimes which would require changing the project settings to match the correct version. 

I don't think this should be too hard to automate using GitHub Actions. If I get some time again soon I'll take a look at that as well.

Quinn

**Reply by atummons** [reply]
Quinn, thanks for the helpful write up.  I had to do a bit of googling because the linking was less intuitive than I would have imagined.  I did get a successful build.  I named my project P3ROC and its output is in my \source\repos\P3ROC\x64\Release folder a file titled P3ROC.dll.  Googling tells me I can just rename the file and change the extension to .pyd (pinproc.cp39-win_amd64.pyd).

Is that actually accurate, or did a I miss a step somewhere that made it have a generic name instead of that specific name?

If its not, I did the linking per your step and changed my values to Release x64 and clicked Play Local Windows Debugger and got 1 build successful.  So if I missed a step or did something out of the norm, let me know.  If that is the right steps, then I will get it added to my local MPF folder and try to run on my hardware.

**Reply by atummons** [reply]
Upon additional comparison, my generated file was 11KB, and the existing 64s in the repo are all 79KB.  So I must be missing the place where I am trying it to the python version.  I am pulling in the lib and file, but maybe it's being ignored?

**Reply by qcapen** [reply]
I double-checked my build and I had the same result (11KB). Just to make sure, I renamed my output file to pinproc.cp39-win_amd64.pyd and imported into Python 3.9:

`>>> import os
>>> os.chdir("D:\Development\Pinball\multimorphic\windows\pypinproc\\x64\Release")
>>> import pinproc
>>> machine_type = pinproc.MachineTypeWPC
>>> pr = pinproc.PinPROC(machine_type)
FT_ListDevices(2)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: FT_ListDevices(2)

>>> print(machine_type)
3
>>> print(pinproc)
<module 'pinproc' from 'D:\\Development\\Pinball\\multimorphic\\windows\\pypinproc\\x64\\Release\\pinproc-cp39-win_amd64.pyd'>
>>>`
It sure looks like it is working (not surprising I received an OSError since I don't actually have any hardware) as the module imported just fine and I was able to call a few functions (the code comes from pypinproctest.py in the project). The old way I built the pypinproc extension library included a static pinproc library linked in the build which probably made it a lot bigger. Now by compiling everything into a single dll, only the code that is actually called is included in the build. Without giving it a more thorough test I don't know for sure that it is working, but my initial tests are promising. I initially received an error that the dll could not be loaded, but I copied the ftd2xx64.dll from the ftd2xx\amd64 subdirectory into the same directory as pinproc.pyd and renamed it ftd2xx.dll (you could also put it in your Windows/System32 folder or anywhere in the path). After doing that I could import it into python.

You'll also have to place the dll into the mpf/platforms/pinproc/windows directory with the correct name (pinproc.cp39-win_amd64.pyd) in order for mpf to load it.

**Reply by atummons** [reply]
Awesome. So sounds like my test matches yours. I'll get my computer hooked up to run on P3ROC hardware, drop that renamed file into the windows folder and see if my flippers still work.

Thanks again, absolutely no way I would have gotten off the ground without your help!

**Reply by atummons** [reply]
That does not appear to be the trick.  I imported the file into the windows folder (I even copied and renamed it to the win32 file in case it expected both.  I am in PyCharm on Python3.9 64 bit using the debug_run_game function for p3_roc and it is not finding the file (or finding it but not being able to import.  Debug puts it in a black hole and you don't see what is happening deeper.  Running through cmd provides the same error: 
'Could not import "pinproc". Either the library is not installed or is missing ' mpf.exceptions.runtime_error.MpfRuntimeError: Runtime Error in P-Roc: Could not import "pinproc". Either the library is not installed or is missing some of its dependencies.

![image](https://github.com/missionpinball/missionpinball/assets/60987561/52b2b7b5-1d80-4daa-a9de-dbea6d0a9ea8)

Note that in PyCharm with the 3.8 interpreter, it works fine, so I don't think its a broken process, but rather not liking that file.  Is there a chance it needs something else beyond what I did?

**Reply by qcapen** [reply]
The one I built is definitely importable in Python. Do you want me to send it to you to try (and the VC++ project)? If so, send me an email (my address is in the mpf AUTHORS file). I can send you a download link.

Quinn

**Reply by atummons** [reply]
First off... everyone that uses multimorphic hardware owes Quinn many many beverages of his choosing!  I was able to get it working with the file Quinn sent me.  I had two computers side by side with the same code, one running 0.55 and python 3.8 and one running 0.56 and python 3.9.  Both were able to connect, run hardware scan, and register switches coils, autofire coils, etc.  So as far as I can test with my current hardware it appears to be working as expected!!!

I will go through to see why I couldn't get it working on my build and get clear documentation written so that this can easily be supported in the future as versions increment.

**Reply by borgdog** [reply]
excellent news! thanks very much @qcapen and @atummons

**Reply by atummons** [reply]
Sorry, work went to hell last week and got tied up.  I was able to get the 64-bit compiled (and matched what Quinn had).  Then got the 32-bit compiled as well (not sure if it's really needed, but to be consistent if people still have 32-bit machines).  I will get those pushed up so that they can be included for others going forward.  I will be doing that shortly.

@qcapen I went ahead and compiled 3.10 and 3.11 off their respective Python while I was at it and can do 3.12 when it comes out in a few months.  Do you see any issue of pushing up those files so that they are there when MPF moves to future versions of Python?

Also, any thoughts about pushing up the VS Project with some instructions?  If you have the project, with my three changes that need made for deployment of a new version for both 32 and 64, it's something that anyone else could pick up at this point until/if its automated.  Would that be a new project all together that houses it?  If so, should that be owned by Mission Pinball?

**Reply by qcapen** [reply]
Glad you got everything working. Feel free to add the 3.10 and 3.11 (and 3.12 when it becomes available) to the mpf project. That should buy quite a bit of time for compatibility with future releases. The eventual goal would be to build the Windows binaries in the Mission Pinball version of the pypinproc repository using GitHub actions. For now the manual build should suffice. I suppose the VS project could be posted in another missionpinball GitHub repo with modification instructions, although if any of the pinproc or libpinproc code changes it would have to be manually updated (I'm not sure if the FTDI source can be included in our repo or whether that would have to be downloaded and added manually to the project depending on the license). Perhaps it is sufficient enough to just post very detailed instructions somewhere to recreate the project.

**Comment by toomanybrians** [comment]
I don't use a P-ROC so I can't comment on whether it would work on Mac or Linux under Python 3.7+, but for your other question, all the old documents are still available. On the docs.missionpinball.org site, in the lower left corner, you can click the green version text. (Probably "v:stable" or "v:latest" in the docs you're looking at.) Then from there you should see links to 0.55 and prior docs versions.

If you're part of the PinDev Slack group (I think there's where P-ROC/P3-ROC support is nowadays), you can try posting there to see if anyone is willing to see why the P-ROC drivers don't compile for Python 3.7+, though I believe most P-ROC/P3-ROC users use PyProcGameHD/SkeletonGame which runs on Python 2.x so you might not have luck there.

Also as for adding a warning to the docs, the MPF docs are also open source, so if you're feeling adventurous and want to make this change to the docs yourself, you can click the "Edit on GitHub" link in the upper right corner of any MPF docs page to make changes to that document page, and if you need more detailed instructions, there's a topic called "Contributing to MPF docs" in the table of contents all the way at the bottom under "About MPF".

**Reply by atummons** [reply]
@toomanybrians I would argue the comment about P-Roc (P3) not using MPF would not be accurate. There seem to be a ton of people using that, and more that I personally know than other "off the shelf" if you exclude OPP.

All that to say the previous maintainers (Jan, Quinn, Anthony) managed and supported all platforms. I know your work and updates to text has made it clear that MPF maintains itself, and all platforms must maintain their plugins. If that's incorrect, please correct me. With that said, has there been any effort to communicate that to the platform hosts? Obviously you handle FAST, but has Gerry/Ocean been made aware?

Thanks!

**Reply by Midway-Man** [reply]
@atummons I've contacted Gerry today, maybe he has a solution. Let's wait and see. ;)

In the meantime I finished working on the CAD of Whitewood #2. Time well spend :D

**Reply by atummons** [reply]
Thanks for doing that.  Fingers crossed!

**Reply by toomanybrians** [reply]
Ah, ok good to know @atummons that there are lots of P-ROC Windows MPF users. Makes sense that I wouldn't know since I don't use a P-ROC or Windows myself. Hopefully we hear back from one of them to update the Windows drivers! 🤞Or maybe they're just using older versions of MPF which are compatible with Python 3.6 or older? Anyway let's see what Gerry comes back with.

Yes, you're right about me asking for individual platforms to maintain their own platform interfaces. (Or for them to find someone to do it.) I'm still looking for someone to own the maintenance for the MPF project in general. I don't have the time or skill that Jan did (the new installers alone took me like 2 months) so that's why I'm trying to set the expectations in the docs that no one is really in charge, so people should set their expectations accordingly. But if there's a better approach, I'm open to it!

**Reply by atummons** [reply]
@toomanybrians all sounds great! I like to contribute as I can and understand the ask, but I am certainly not knowledgeable enough to review code. As long as we have someone who can, even periodically, review community contributions, I think we're in good shape!

**Comment by colemanomartin** [comment]
Bump.  It's not just windows, its Ubuntu also.  I wonder about Mac.

**Reply by Midway-Man** [reply]
Bummer… Thanks for sharing this info. 
I was thinking about moving my system over to Linux. Well seems like this won’t solve the issue.

Unfortunately I haven’t got any reply from Multimorphic. Will try another approach.

**Comment by colemanomartin** [comment]
I was able to get it working under ubuntu 20 today.  In my case there were a couple hangups.  First I went with ubuntu 20 because it defaults to Python 3.8.  Then I ran the installer scripts one line at a time from the command line.  I ran into a minor problem with a kivi version but was able to set it to the correct version when it initially failed.  I believe I used the suffix for version ==0.1.1 Then I had to alter the last command of the install-proc script from sudo python3 setup.py install  to python3 setup.py install --user  save the script give it the sudo chmod + x and run it.  After that I had MPF connecting to Multimorphic hardware.  Given my success, I think I can also get it to work with Ubuntu 22 by setting up a python virtual environment for python 3.8.  But that's a problem for another day.

**Reply by Midway-Man** [reply]
You are using Python 3.8? Which version of MPF do you use? 
Didn’t the instructions say MPF 0.56 will only work with Python 3.9?

**Comment by atummons** [comment]
3.9 is only a requirement on Windows. Brian or someone else in the know would have to explain why. I looked at changes from 3.8 to 3.9 and don’t see any dependencies that require it, but I’m certain I just don’t fully understand it.Sent from my iPhoneOn Apr 26, 2023, at 10:25 AM, Midway-Man ***@***.***> wrote:﻿
You are using Python 3.8? Which version of MPF do you use?
Didn’t the instructions say MPF 0.56 will only work with Python 3.9?

**Comment by toomanybrians** [comment]
Files have been added, nice work everyone.
https://github.com/missionpinball/mpf/pull/1706
